aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+                // EXECUTING (finish_execution sets Executed status and DEFER or SAFE flag).
+                self.swap_dependency_flag_any(
+                    &[dependency_flags::DEFER, dependency_flags::SAFE],
+                    dependency_flags::SAFE,
+                    "remove_stall",
+                )?;
+            }
+
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Checks if an incarnation has already been marked for abortion.
+    ///
+    /// This can be called during an ongoing execution to determine if the
+    /// execution has been concurrently aborted. This allows the executor
+    /// to return early and to discard the results.
+    pub(crate) fn already_try_aborted(&self, incarnation: Incarnation) -> bool {
+        self.next_incarnation_to_try_abort.load(Ordering::Relaxed) > incarnation
+    }
+
+    /// Checks that the shortcut flag is SAFE. This is used by the scheduler
+    /// when removing a previously propagated stall signal, and it is safe to
+    /// use the shortcutbecause of the best-effort nature of the stall mechanism.
+    pub(crate) fn shortcut_executed_and_not_stalled(&self) -> bool {
+        matches!(
+            self.dependency_shortcut.load(Ordering::Relaxed),
+            dependency_flags::SAFE
+        )
+    }
+
+    /// Checks if the transaction is ready for scheduling and not stalled.
+    /// This is used to determine if a transaction should be considered
+    /// for execution by the scheduler.
+    pub(crate) fn pending_scheduling_and_not_stalled(&self) -> bool {
+        let guard = self.inner_status.lock();
+        guard.pending_scheduling().is_some() && self.num_stalls.load(Ordering::Relaxed) == 0
+    }
+
+    // === More inner status wrappers ===
+
+    pub(crate) fn is_executed(&self) -> bool {
+        self.inner_status.lock().is_executed()
+    }
+
+    pub(crate) fn ever_executed(&self) -> bool {
+        self.inner_status.lock().ever_executed()
+    }
+
+    pub(crate) fn last_incarnation(&self) -> Incarnation {
+        self.inner_status.lock().incarnation()
+    }
+
+    /// Gets the current status of this transaction.
+    pub(crate) fn status(&self) -> StatusEnum {
+        self.inner_status.lock().status.clone()
+    }
+}
+
+// Private interfaces.
+impl ExecutionStatus {
+    /// Performs an atomic swap operation on the dependency shortcut flag and validates
+    /// that the previous value matches one of the expected values.
+    ///
+    /// # Parameters
+    /// - `expected_values`: Array of possible expected current flag values
+    /// - `new_value`: The new flag value to set
+    /// - `context`: A string describing the context for error messages
+    ///
+    /// # Returns
+    /// - `Ok(prev)` if the swap was successful, returning the previous value
+    /// - `Err(PanicError)` if the previous value didn't match any expected value
+    fn swap_dependency_flag_any(
```

Comment:
This is not test-and-set? The value will be swapped anyway, even when the previous value does not match the expected value. Is this the right behavior?

Reply:
good question! Here, since it must be one of the expected values, we could even assert, but we have a better way to return a PanicError and possibly recover instead of crashing if we did assert. So in other words, we don't even need to swap, we can just update, we only swap to check the invariant, and PanicError definitely stops the transaction, maybe the whole block execution, to retry some fallbacks, so whether we leave it swapped isn't important.

Funny aside, test-and-set because it's on the bits, is often defined as "check and if false set to true (return true)", but in reality is often implemented by "set to true, return true if previous value was false".